### ISSUE TITLE
Native libraries that need to be gemified

### ISSUE THREAD

**Issue by headius** [jruby/jruby#6682]
There are still several native libraries that CRuby has moved into gems and we have not. We need to deal with these so people can add those native gems to dependencies without issue. See #4875 for earlier discussion about gemification.

The remaining missing gems are summarized below:

bigdecimal - we still have our own version and it is not integrated into the gem in any way. The gem maintainer has expressed some reluctance to incorporating JRuby code in the past.
date - we have our own impl for 90% of this and it has not been incorporated into the gem.
dbm, gdbm, and sdbm have never been supported and likely never will.
digest - we have our own impl and it has not been incorporated into the gem.
etc - we have our own impl and it has not been incorporated into the gem. This could potentially be replaced by Ruby FFI code and made easier to add to the gem.
fcntl - we have our own impl and it has not been incorporated into the gem. This could also be replaced by FFI code.
fiddle - we have a pure-Ruby impl based on FFI that has not been incorporated into the gem. In the past we had talked with @chrisseaton about unifying on a shared fiddle impl that uses FFI, or else a shared FFI based on fiddle, but that has not happened.
openssl - we release this in gem form but our version of the library has not been incoporated into the standard gem.
stringio - we have our own impl and it has not been incorporated into the gem.
strscan - we have our own impl and it has not been incorporated into the gem (ruby/strscan#15).
zlib - we have our own impl and it has not been incorporated into the gem.

**Comment by hsbt** [comment]
👍  I'll support this.

**Comment by headius** [comment]
strscan gemification is started in https://github.com/ruby/strscan/pull/25.

**Comment by headius** [comment]
See ruby/digest#20 / ruby/digest#21 for the gemification of Digest.

**Comment by headius** [comment]
stringio started in https://github.com/ruby/stringio/pull/21

**Comment by hsbt** [comment]
@headius Thanks for your work! Did we get the approval for relicensing JRuby libraries that are this work at https://github.com/ruby/digest/pull/21#issuecomment-861015401 ?

I'm not sure what we should get the approval for that each pull requests.

**Comment by headius** [comment]
@hsbt I believe we got approval from everyone except @chrisseaton in that PR and updated my comment to be a checklist.

@chrisseaton You responded affirmatively to the PR, but could you add something along the lines of "I consent to having my contributions relicensed etc etc"?

**Comment by headius** [comment]
@hsbt FWIW having talked with legal-ish folks about license changes, as long as everyone publicly states that they are ok with it that is generally enough, legally-speaking.

**Comment by chrisseaton** [comment]
https://github.com/ruby/digest/pull/21#issuecomment-912466365

**Comment by headius** [comment]
@chrisseaton Thank you!

@hsbt That's everyone!

**Comment by headius** [comment]
I will try to make sure the other PRs can get relicensed as well.

**Comment by headius** [comment]
I have updated the description with a complete list of Ruby 3.0 gemified libraries and links to current issues and PRs.

**Comment by headius** [comment]
> complete list

At least a complete list of the gemified libraries that we could not trivially incorporate. Several like `fileutils` and `rdoc` already worked find and have just been switched to default gems.

**Comment by headius** [comment]
The io-wait, stringio, and strscan gems now support JRuby since the last time I updated this.

**Comment by headius** [comment]
bigdecimal and openssl have merged JRuby gem configs for stub gems that depend on our own libraries:

* https://github.com/ruby/bigdecimal/pull/250
* https://github.com/ruby/openssl/pull/598

These have not yet been released in gems.

**Comment by headius** [comment]
@adurgin Good catch! I'll see about getting a PR into net-ftp to make it work on JRuby.

**Comment by headius** [comment]
@adurgin I've submitted a PR to net-ftp so we're just waiting for that to merge and release.

https://github.com/ruby/net-ftp/pull/38

**Comment by adurgin** [comment]
Fantastic. Thank you.

For anyone else affected, I was able to work around it in the interim with a monkey patch of initialize.

In case one is better than the other, I did notice that net-http approached it in a slightly different way: https://github.com/ruby/net-http/blob/master/lib/net/http.rb#L1681

**Comment by headius** [comment]
net-ftp 0.3.7 has been released with a fix! I'll update JRuby to ship that where appropriate and we should be good to go.

**Comment by headius** [comment]
net-ftp updated in https://github.com/jruby/jruby/pull/8305

This will merge into 9.4 and 10 once it's green.
